### DISCUSSION TITLE
Feature request: Support vault aws auth backend

### DISCUSSION THREAD

**Discussion by jmcarp** [concourse/concourse#1367] [Ideas]
## What challenge are you facing?

It would be useful if atc could authenticate to vault using the [aws backend](https://www.vaultproject.io/docs/auth/aws.html). It looks like the current vault config allows users to pass arbitrary key-value options to arbitrary auth backends, but the aws backend requires some information that isn't available when atc starts. Specifically, the [ec2 auth method](https://www.vaultproject.io/docs/auth/aws.html#ec2-authentication-method) requires making a call to the aws metadata endpoint to get the signature of the instance metadata to pass it to vault, and the [iam auth method](https://www.vaultproject.io/docs/auth/aws.html#iam-authentication-method) requires making a call to `sts:GetCallerIdentity` (the vault cli implementation lives at https://github.com/hashicorp/vault/blob/da2c918195ebae40495b8db88d245f31948d5b31/builtin/credential/aws/cli.go#L19-L100).

It looks like supporting these auth methods would require teaching the atc to make the appropriate calls to the aws metadata endpoint or api on login and renew actions if the relevant backend is selected. If that makes sense, I'd be happy to work on a patch.

**Comment by jtarchie** [comment]
Thanks for the issue @jmcarp and doing some of the research. I'm going to discuss this a little more with the team.

**Comment by willejs** [comment]
@jtarchie This would be a really great feature! Currently using IAM roles in concourse is mostly hacky and insecure. Proper AWS backend support for vault would mean much improved security!

**Comment by jmcarp** [comment]
WDYT @jtarchie @chendrix @vito ? Would you be interested in implementing this feature, or in reviewing a patch?

**Comment by mogul** [comment]
@jtarchie @chendrix @vito have you had any chance to think about this? We're ready to start making some decisions about how we handle secrets, but the Concourse direction/receptivity here is a blocker for us.

**Comment by novas0x2a** [comment]
For what it's worth, I think this already works, kinda? I launch concourse with a wrapper that just does: 
```
export CONCOURSE_VAULT_CLIENT_TOKEN="$(vault auth -token-only ${COMPANYNAME_VAULT_AUTH_ARGS})"
exec /sbin/concourse "$@"
```
and `COMPANYNAME_VAULT_AUTH_ARGS` is set to `-method=aws role=my-cool-role` in the conf where I set all the other env vars for concourse. You end up getting the same token you'd get anyway if concourse had direct support.

Direct support would be cleaner, but I wouldn't consider lack of support for it to be a blocker. A note on the above: small clarification on the IAM auth method. The steps are:
1) Create a [GetCallerIdentity](https://github.com/hashicorp/vault/blob/da2c918195ebae40495b8db88d245f31948d5b31/builtin/credential/aws/cli.go#L60) request object. (Don't send it).
2) [Sign it](https://github.com/hashicorp/vault/blob/da2c918195ebae40495b8db88d245f31948d5b31/builtin/credential/aws/cli.go#L66) with the configured aws creds
3) [Extract](https://github.com/hashicorp/vault/blob/da2c918195ebae40495b8db88d245f31948d5b31/builtin/credential/aws/cli.go#L69) all of the pieces of the still-unsent request (method/url/headers/body).
4) [Pass](https://github.com/hashicorp/vault/blob/da2c918195ebae40495b8db88d245f31948d5b31/builtin/credential/aws/cli.go#L84) the request data to vault, who then makes the api call on your behalf.

I should note that the code for the above (as of a more recent vault client version) actually lives in an extracted and exported [go function](https://github.com/hashicorp/vault/blob/79b63deaf50c216e8b1d5edd12388ad85d137537/builtin/credential/aws/cli.go#L21) so concourse could be spared some of the nitty gritty details of the above.

**Comment by iuriaranda** [comment]
Is there any update on this?

As of now, we're using the approach mentioned by @novas0x2a as a workaround, but I don't think it can be considered a final solution to this issue. The main downside of such approach is that Concourse is not able to fetch a new token when the old one reaches [its maximum TTL](https://www.vaultproject.io/docs/concepts/tokens.html#the-general-case). In that case, the Vault integration breaks, until the process is restarted and the wrapping script runs again to fetch a new token.

**Comment by digitalstoic** [comment]
This looks like a great feature to have! I am also interested to find out if you have any updates on this.

**Comment by dvianello** [comment]
Following @novas0x2a advice, I ended up building a small golang program wrapping around that go function. 

In this way, we can set `CONCOURSE_VAULT_AUTH_BACKEND` to `aws` and pass the sts call into `CONCOURSE_VAULT_AUTH_PARAM`. Concourse is then able to auth against vault, get a token and since we're issuing a `periodic` token will live happily ever after as long as it asks to renew it within the configured `period`. It has been running on two boxes for a while now, and no issues so far. 

Script lives [here](https://github.com/dvianello/vault-iam-request) should anybody be interested!

**Comment by PavelMikhailouski** [comment]
@jtarchie @chendrix @vito any updates on the item? 

Would be an awesome feature to have out of the box!

**Comment by ykhemani** [comment]
Hi! Just wanted to bubble this up and see it may be possible to get this feature request prioritized, @jtarchie @chendrix @vito 

Thanks so much!

**Comment by vito** [comment]
Sorry for the lack of updates on this, there's a lot to wrangle with IAM/AWS/Vault and I've been focusing on issues that somewhat related (#5663, #3023, #5813) which had me fooled into believing I was indirectly working on this too, but now that I look at it this case is a bit more cut and dry.

@ykhemani We would be more than happy to review a PR that uses the [`GenerateLoginData` method](https://github.com/hashicorp/vault/blob/b557b76ac16d7da6ac049b53b6867aec46861531/builtin/credential/aws/cli.go#L38) that @novas0x2a linked previously! Such a PR should also be able to re-authenticate when needed, per @iuriaranda's comment.

We don't really have the bandwidth to prioritize this above the [v10 roadmap](https://blog.concourse-ci.org/core-roadmap-towards-v10/) or K8s runtime initiatives, so I'm going to convert this to a Discussion so it's more obvious that this isn't something on our roadmap at the moment (per our [issue process](https://blog.concourse-ci.org/nobody-wants-a-stale-bot/)).

Thanks everyone for your patience.
